Automattic/jetpack
# autoloader: Preload classes on plugin upgrade to avoid mid-upgrade fatals

#### Changes proposed in this Pull Request:

It's possible that the new version of the plugin will move some files
around that some later hook will try to load, causing a fatal as the
autoloader will still be looking in the old path. To avoid this, have
the Jetpack autoloader detect when an upgrade is happening and pre-load
all the classes in the plugin.

We do the same for plugin removal, for similar reasons.

Note this doesn't work for classes loaded via PSR-0 or PSR-4. But that
shouldn't matter for us since we always generate an optimized (classmap)
autoloader for releases of our plugins.

Also of note is that this won't work if you're manually installing a
specific version of the plugin as `wp plugin install jetpack --version=10.5`,
as wp-cli does that as an "install" rather than an "upgrade" operation.

#### Jetpack product discussion


p9dueE-4c4-p2

#### Does this pull request change what data or activity we track or use?


No

#### Testing instructions:






* Create a fresh site, e.g. with JN.
* Use Jetpack Beta to downgrade Jetpack to 10.4, and to install the release version of itself. Uninstall any other monorepo plugins. Also delete any `wp-content/plugins/*-dev`, just to be sure.
* Connect Jetpack (sync won't try to sync post-upgrade if you don't).
* Add the following mu-plugin:
  ```php
  <?php
  add_filter( 'pre_set_site_transient_update_plugins', function ( $transient ) {
  	if ( isset( $transient->response['jetpack/jetpack.php'] ) ) {
  		$transient->response['jetpack/jetpack.php']->new_version = '10.5';
  		$transient->response['jetpack/jetpack.php']->package = "https://downloads.wordpress.org/plugin/jetpack.10.5.zip";
  	}
  	return $transient;
  } );
  ```
* At this point the plugins page and `wp plugin upgrade` should both want to upgrade Jetpack to 10.5. Doing so should give a mid-upgrade fatal as described in p9dueE-4c4-p2.
* Re-install Jetpack 10.4.
* Use Jetpack Beta to check out the version of Jetpack Beta from this PR.
  * Optional: `wp shell <<<'spl_autoload_functions()[0];'` should indicate that it's using `Automattic\Jetpack\Autoloader\jp567fa3f555de8fd218dfdc1688bb97b5_betaⓥ3_1_1_alpha\PHP_Autoloader`.
* Now upgrading via the plugins page or `wp plugin upgrade` should succeed without a mid-upgrade fatal.
